### PR TITLE
Add subnodes to introspect result

### PR DESCRIPTION
**Pull request by goldenice**
Needed to access all subnodes of a certain node, and I noticed this module doesn't offer that functionality, although I may have overlooked something.

Anyhow, I decided to code it up, and it was surprisingly simple. Perhaps this would be useful upstream as well?

I tried to match the coding style as closely as possible.

### CODE REVIEW

**Comment by sidorares** [comment]
Thanks @goldenice !

Can you provide xml example (maybe even put it under test fixtures + add introspection integration tests)? I thought this was handled before, but looks like not

**Comment by goldenice** [comment]
@sidorares Yeah sure, I'll provide the xml sample later today, and I'll see what I can do about the tests :)

**Comment by goldenice** [comment]
This was the XML I needed it for. The specific sample is from the Bluez5 API, every subnode in this XML signifies a subdevice.

The path here was `org/bluez/hci0`, and every subnode is `/org/bluez/hci0/[node name]`, as expected.

```
<!DOCTYPE node PUBLIC "-//freedesktop//DTD D-BUS Object Introspection 1.0//EN"
"http://www.freedesktop.org/standards/dbus/1.0/introspect.dtd">
<node><interface name="org.freedesktop.DBus.Introspectable"><method name="Introspect"><arg name="xml" type="s" direction="out"/>
</method></interface><interface name="org.bluez.Adapter1"><method name="StartDiscovery"></method><method name="StopDiscovery"></method><method name="RemoveDevice"><arg name="device" type="o" direction="in"/>
</method><property name="Address" type="s" access="read"></property><property name="Name" type="s" access="read"></property><property name="Alias" type="s" access="readwrite"></property><property name="Class" type="u" access="read"></property><property name="Powered" type="b" access="readwrite"></property><property name="Discoverable" type="b" access="readwrite"></property><property name="DiscoverableTimeout" type="u" access="readwrite"></property><property name="Pairable" type="b" access="readwrite"></property><property name="PairableTimeout" type="u" access="readwrite"></property><property name="Discovering" type="b" access="read"></property><property name="UUIDs" type="as" access="read"></property><property name="Modalias" type="s" access="read"></property></interface><interface name="org.freedesktop.DBus.Properties"><method name="Get"><arg name="interface" type="s" direction="in"/>
<arg name="name" type="s" direction="in"/>
<arg name="value" type="v" direction="out"/>
</method><method name="Set"><arg name="interface" type="s" direction="in"/>
<arg name="name" type="s" direction="in"/>
<arg name="value" type="v" direction="in"/>
</method><method name="GetAll"><arg name="interface" type="s" direction="in"/>
<arg name="properties" type="a{sv}" direction="out"/>
</method><signal name="PropertiesChanged"><arg name="interface" type="s"/>
<arg name="changed_properties" type="a{sv}"/>
<arg name="invalidated_properties" type="as"/>
</signal>
</interface><interface name="org.bluez.Media1"><method name="RegisterEndpoint"><arg name="endpoint" type="o" direction="in"/>
<arg name="properties" type="a{sv}" direction="in"/>
</method><method name="UnregisterEndpoint"><arg name="endpoint" type="o" direction="in"/>
</method><method name="RegisterPlayer"><arg name="player" type="o" direction="in"/>
<arg name="properties" type="a{sv}" direction="in"/>
</method><method name="UnregisterPlayer"><arg name="player" type="o" direction="in"/>
</method></interface><interface name="org.bluez.CyclingSpeedManager1"><method name="RegisterWatcher"><arg name="agent" type="o" direction="in"/>
</method><method name="UnregisterWatcher"><arg name="agent" type="o" direction="in"/>
</method></interface><interface name="org.bluez.HeartRateManager1"><method name="RegisterWatcher"><arg name="agent" type="o" direction="in"/>
</method><method name="UnregisterWatcher"><arg name="agent" type="o" direction="in"/>
</method></interface><interface name="org.bluez.ThermometerManager1"><method name="RegisterWatcher"><arg name="agent" type="o" direction="in"/>
</method><method name="UnregisterWatcher"><arg name="agent" type="o" direction="in"/>
</method><method name="EnableIntermediateMeasurement"><arg name="agent" type="o" direction="in"/>
</method><method name="DisableIntermediateMeasurement"><arg name="agent" type="o" direction="in"/>
</method></interface><node name="dev_24_0A_64_51_49_9A"/><node name="dev_4A_38_76_54_3F_6A"/><node name="dev_78_F8_82_47_DF_68"/><node name="dev_7C_D1_C3_34_F5_CC"/><node name="dev_9C_20_7B_E8_1F_8B"/><node name="dev_D0_17_D6_03_E4_65"/><node name="dev_F4_5C_89_8D_37_DA"/></node>
```

The array in object.nodes using this XML input was:

```
[ 'dev_4A_38_76_54_3F_6A',
  'dev_78_F8_82_47_DF_68',
  'dev_7C_D1_C3_34_F5_CC',
  'dev_9C_20_7B_E8_1F_8B',
  'dev_D0_17_D6_03_E4_65',
  'dev_F4_5C_89_8D_37_DA' ]
```

as expected.

File: lib/introspect.js [hunk 1]
```diff
@@ -13,8 +13,16 @@ module.exports = function(obj, callback) {
...
           if (!result.interface) return callback(new Error('No such interface found'));
           var proxy = {};
+          var nodes = [];
           var i, m, s, ifaceName, method, property, signal, iface, a, arg, signature, currentIface;
           var ifaces = result['interface'];
+          var xmlnodes = result['node'];
+          if (xmlnodes == null) xmlnodes = [];
+
+          for (i = 1; i < xmlnodes.length; ++i) {       // Start at 1 because we want to skip the root node
+              nodes.push(xmlnodes[i]['@']['name']);
```
Comment (sidorares) [inline lib/introspect.js]: what if there is no `xmlnodes[i]['@']` element?

Comment (goldenice) [inline lib/introspect.js]: I guess it'll break. Then again, a few lines later you do the same thing with `iface['@'].name`, so I didn't assume I'd have to put a check on that.

**Comment by jrobeson** [comment]
@goldenice I've been working on the dbus bluez api for https://github.com/sandeepmistry/noble but I haven't really gotten very far with introspection due to this issue (and lack of time). Have you gotten any farther with this? or have any code to share?

**Comment by goldenice** [comment]
@jrobeson Well, the code in this pull request is currently in production at the company I work at (Athom, http://athom.com), and it seems to work just fine. Then again, we're not heavy users of introspection in DBus (just one call to fetch all currently connected bluetooth devices) so I can't really say if this code would work for every use case.

**Pull request merged**